File: shinjuku_hs/2016-10-30/adt_recursive.rb
class Matryoshka
  def initialize(matroyoshka = nil)
    @matroyoshka = matroyoshka
  end

  def matroyoshka
    @matroyoshka
  end
end

t = Matryoshka.new()
t1 = Matryoshka.new(t)
t2 = Matryoshka.new(t1)

def search(mat)
  return [] unless mat
  [mat, search(mat.matroyoshka)]
end

def count(mat)
  search(mat).flatten.count
end

puts count(t2)

File: shinjuku_hs/2016-10-30/adt_recursive.hs
import Data.List
import Data.Either

data Matryoshka = Nest Matryoshka | Term deriving Show

search :: Matryoshka -> [Matryoshka]
search (Nest a) = [Term] ++ search a
search Term = [Term]

count :: Matryoshka -> Int
count = length . search

main = do
  print $ search Term
  print $ search (Nest Term)
  print $ search (Nest (Nest Term))
  print $ count (Nest (Nest Term))
  print $ count (Nest (Nest (Nest Term)))
